### ISSUE TITLE
'no_or_const_delay_mode' is not set to True for no or scalar delays

### ISSUE THREAD

**Issue by denisalevi** [brian-team/brian2cuda#24]
Testing for no or scalar delays is not implemented correctly. In standalone mode the delay parameter and therefore its `.scalar` and `.size` flags are not accessible before simulation hasn't been run. At least not in the way it was intended in the current version [here in commit bac8b5a](https://github.com/brian-team/brian2cuda/blob/bac8b5a/brian2cuda/device.py#L92).

I would leave this for now and look into it once the `no_or_const_delay_mode` implementation is finished and we updated to brian2 master (#22).

some thoughts:
- as far as I remember string expressions such as `delay = '5 * ms * rand()'` or `delay = 'i * ms'` get translated into c++/cuda code in standalone mode and thats why their values are not available before simulation
- check when the `delay.scalar` and `delay.size` flags are updated
- if they are updated after simulation, maybe some checking for appearances of `rand()`, `i`, `xi` etc. in the delay expression might be necessary in order to determine weather there will be different delays for different synapses or not (check available brian2 functions for string characterisation)

Until this is fixed, `no_or_const_delay_mode = False` is hard coded [here](https://github.com/brian-team/brian2cuda/blob/bac8b5a/brian2cuda/device.py#L97)!

**Comment by mstimberg** [comment]
Hi, just to clear up some (understandable...) confusion here:
The `constant`, `scalar` and `size` attributes are attributes of the `Variable` object which is defined on the Python level (it is the basis of the code generation, together with the information about types, units, etc.) -- those *are* available before the run, as they are not supposed to change. `size` is an exception, as it is a dynamic variable that can be resized during the run.

There are two ways to specify a synaptic delay: either you set it after the creation of the `Synapses` object with something like `synapses.delay = ...` or you can set it directly in the `Synapses` initializer call:
```Python
synapses = Synapses(..., delay=5*ms)
```
In the first case (`synapses.delay = ...`), the `Variable` object has `scalar==False` and one delay value per synapse is stored (it would be the same value for all synapses after something like `synapses.delay = 5*ms`, though). In the second case (`Synapses(..., delay=...)`), the `Variable` object has `scalar==True` and stores a single value for all synapses.
I think this is the only distinction that you'll have to do: if the user specifies the delay with the second syntax, you'll have a single delay for all synapses, otherwise you'll have a (potentially) different delay for the synapses. This is the fundamental difference on the memory level.

Now, you might *additionally* want to optimise for the (rather common...) case that the user did not use the constant-delay syntax in the `Synapses` initializer syntax but either directly specified the same delay for all synapses or this just accidentally happened to be the case (e.g. random delays with a small jitter, but all delays are the same after rounding to `dt`). Don't do this on the Python level (you couldn't catch this for the second case with small random deviations) but rather on the C++/CUDA level. We have done this optimisation in standard Brian2 as well, it is in the `prepare` method of the `SpikeQueue`: https://github.com/brian-team/brian2/blob/master/brian2/synapses/cspikequeue.cpp#L48
During the conversion of the synaptic delay float values to integer time steps we note down the minimum and maximum delay; if both are identical we set `scalar_delay` to `true` and then use a different code path in the `push` method: https://github.com/brian-team/brian2/blob/master/brian2/synapses/cspikequeue.cpp#L151

**Comment by denisalevi** [comment]
Ah, ok. Thank you for clarification! I was confused that specifying only a single delay through the `synapses.delay = 1*ms` syntax resulted in `scalar==False`. 

Currently the `no_or_const_delay_mode` flag is set on the Python level and is used in the template macros to decide what code to paste into the `.cu` files. This needs to be changed then to decide on the C++/CUDA level which code path to use instead if we want to check for the case where the same delay is specified for each synapse through `synapses.delay =` syntax or with small jitter.

**Comment by mstimberg** [comment]
> This needs to be changed then to decide on the C++/CUDA level which code path to use instead if we want to check for the case where the same delay is specified for each synapse through synapses.delay = syntax or with small jitter.

Yes, but I would not worry about this for now. It is only an optimization and the user always has the choice to actively set a scalar delay.

**Comment by denisalevi** [comment]
Yes, I will leave that optimization for later. I have implemented the check for `scalar` for now.

When only checking for `scalar` though, we don't catch the case where there is no delay specified at all (in that case we still have `scalar==False`). The user needs to specify `Synapses(..., delay=0*ms)` in order to have no delays but still use the `no_or_const_delay_mode`.

**Comment by denisalevi** [comment]
I just implemented the C++ side check if all delays are the same (after rounding etc). This works fine and also catches the case, where no delay is specified at all. But its rather unnecessary work, because currently my generated code ends up resizing the delay vector to `num_synapses`, which initializes it with `0`s, and then it realizes that all values are the same and turns on the `scalar_delay` flag.
@mstimberg Is there a way to catch from Python side when the user doesn't specify any delays at all? I digged a little around, but couldn't really find anything.

**Comment by mstimberg** [comment]
> @mstimberg Is there a way to catch from Python side when the user doesn't specify any delays at all? I digged a little around, but couldn't really find anything.

There's unfortunately no elegant way: if the user does not set a scalar delay in the `Synapses` initializer, the delay is considered to be variable, so Brian creates a dynamic array for it. I think the only way (without changes in Brian itself) would be to manually special-case the `delay` variable. Maybe have a look at `brian2genn`: it prevents assignments to the delay variable (because it does not support variable delays), you could maybe use the same approach to check whether the user has set it?

The relevant `brian2genn` code is here:  https://github.com/brian-team/brian2genn/blob/3f710bbb1d32e309ffdcc62e9dd2634cf0f84a6d/brian2genn/device.py#L353

It's a bit more straightforward in `brian2genn`, though, because there we also lack support for multiple pathways.

**Comment by denisalevi** [comment]
This issue is long solved. I'll open another issue for getting rid of unnecessary delay checking in the case where no delay is specified.
